.changes/store-auto-save.md

```diff
@@ -0,0 +1,7 @@
+---
+"store": patch
+---
+
+Add a setting `auto_save` to enable a store to debounce save on modification (on calls like set, clear, delete, reset)
+
+**Breaking change**: `with_store` now takes one more parameter `auto_save: Option<Duration>`
```

Comment:
It feels like `with_store` is growing its original need and requiring too many paramters is not a good DX and will confuse users. I never like this function and I think it should only be used internally. We should look into replacing it with an extension trait, similar to `app.updater_builder()` and other plugins.

Reply:
An app handle extension would be nice, do you have any idea about what that API might be?

Since we have more options about constructing the store now, maybe separate two different functions one being `create_or_load_store` which init the store with settings and put them into the app's managed state and another one being `with_store` which doesn't init the store but simply lock the collection and give the function access to the requested store

Reply:
I honestly don't have an idea in mind now, but don't be afraid to break the whole API design if you have a better API.

Reply:
rewriting it from scratch is totally fine too (if it makes sense). Since this is a v1 plugin it is _not_ on track for a quick stable release (with audit etc).

Reply:
I'm thinking about this

```rust
pub trait StoreExt<R: Runtime> {
    fn load_or_get_store(
        &self,
        path: &Path,
        auto_save: Option<Duration>,
    ) -> Result<StoreWrapper<R>>;
    fn with_store<T>(&self, path: &Path, f: impl FnOnce(&mut Store<R>) -> Result<T>) -> Result<T>;
}
```

`load_or_get_store` loads the store (fallback to empty store if can't load) if it's not in `app.state`'s store collection already

`StoreWrapper` stores the path and forward all the functions like `set`, `delete`, ... using `with_store` internally, and `with_store` doesn't create a new store, everything else says the same as before

and change the js API to a match the rust api (keep the `Store` but change it to be a `Resource` and expose an function like `loadOrGetStore`

Reply:
This is a good direction IMO, but I think we can:
- Remove `with_store`
- Disable creating `Store` manually, and can only be created through the trait ext:
	```rs
	pub trait StoreExt<R: Runtime> {
	  /// get or create a new store
	  fn store(&self, path: &Path) -> Result<Store<R>>;
	}
	``` 
- Move autosave functionality to be a method setter on `Store`

Reply:
This will also fix issues on mobile when users directly call `Store::new` and don't initialize the mobile plugin handle.

Reply:
If we disable creating `Store` manually (which currently can only be created through `StoreBuilder`), we'll probably need to have a `store_builder` trait method to allow more customization

I'm thinking we should just always manage stores in our state and the user can't create a new store if that path is already in our managed stores, but I don't know how we gonna handle those mutexes to be honest, and in that case we might need a `close` function to close the store if the user don't want it to stay in the memory

Also do you think we should do it in this PR or in another one in the future?

Reply:
I think we should just go ahead and implement the refactor since this PR feature already requires breaking changes.

Reply:
> If we disable creating `Store` manually (which currently can only be created through `StoreBuilder`), we'll probably need to have a `store_builder` trait method to allow more customization

We can have another method to create a store builder
```rs
 pub trait StoreExt<R: Runtime> {
   /// get or create a new store
   fn store(&self, path: &Path) -> Result<Store<R>>;
   /// create a store builder
   fn store_builder(&self, path: &Path) -> Result<StoreBuilder<R>>;
 }
```

and similar to `Store`, we shouldn't allow creating `StoreBuilder` manually and prefill it with needed information like mobile plugin handle, so when user eventually call `store_builder.build()` it will behave correctly. This is similar to updater plugin `app.updater_builder()` method.

Reply:
> I think we should just go ahead and implement the refactor since this PR feature already requires breaking changes.

kk, need to think about the new design more then

> This is similar to updater plugin app.updater_builder() method

Sounds good to me

The problem I have right now is all about managed store and non-managed stores, and the mutexes to be honest

Reply:
I would keep the performance optimizations in a later PR as I don't see a way around it for now.

Reply:
The best I can currently think of would probably be separate the creation and the access, and the store is always managed by our store collection, kinda find it end up being just https://github.com/tauri-apps/plugins-workspace/pull/1550#discussion_r1678687871 but replace `load_or_get_store` with `store` and `store_builder` (and maybe the store wrapper itself has a `with_store` function instead of forward all the store methods), also probably a function to remove the store from the managed state

Reply:
Sure, we can have a remove store from managed state function.

Reply:
Actually, `store_builder` kinda doesn't make sense if it returns an existing one with different settings, this is also kinda challenging for lazy loaded stores like what we do in the js side

Reply:
I'm just thinking, what about the stores are in the states and are unique, but they'll only have the data, not the settings, and we can have store wrappers with all the settings like (de)serialize function and auto save, and the public facing APIs will be only store wrappers